cdap/resource_streaming_program.go

```diff
@@ -0,0 +1,248 @@
+			time.Sleep(10 * time.Second)
+		case "STOPPED":
+			log.Println("program still in STOPPED state, waiting 10 seconds. This may occur when redeploying a previously deployed pipeline.")
+			time.Sleep(10 * time.Second)
+		case "RUNNING":
+			log.Println("program successfully reached RUNNING state.")
+			d.SetId(d.Get("app").(string))
+			return nil
+		case "FAILED":
+			return fmt.Errorf("failed to start program in app: %s in state: %s.", d.Get("app"), p.Status)
+		default:
+			log.Println("program still in STARTING state, waiting 10 seconds.")
+			time.Sleep(10 * time.Second)
+		}
+	}
+
+	return nil
+}
+
+func resourceStreamingProgramRead(d *schema.ResourceData, m interface{}) error {
+	return nil
+}
+
+func getProgramStatus(config *Config, statusAddr string) (p ProgramStatus, err error) {
+	req, err := http.NewRequest(http.MethodGet, statusAddr, nil)
+	if err != nil {
+		return
+	}
+
+	b, err := httpCall(config.httpClient, req)
+	if err != nil {
+		return
+	}
+
+	if err := json.Unmarshal(b, &p); err != nil {
+		return p, err
+	}
+
+	return p, nil
+}
+
+func stopProgram(config *Config, stopAddr string) error {
+	req, err := http.NewRequest(http.MethodPost, stopAddr, nil)
+	if err != nil {
+		return err
+	}
+	_, err = httpCall(config.httpClient, req)
+	return err
+}
+
+func resourceStreamingProgramDelete(d *schema.ResourceData, m interface{}) error {
+	config := m.(*Config)
+	name := d.Get("name").(string)
+
+	addr := urlJoin(
+		config.host,
+		"/v3/namespaces", d.Get("namespace").(string),
+		"/apps", d.Get("app").(string), d.Get("type").(string),
+		name)
+
+	// Check status to handle scenarios like a program that is in STOPPING state.
+	statusAddr := urlJoin(
+		addr, "/status")
+	stopAddr := urlJoin(
+		addr, "/stop")
+
+	// Poll until actually reaches STOPPED state.
+	for {
+		p, err := getProgramStatus(config, statusAddr)
+		if err != nil {
+			return err
+		}
+
+		switch s := p.Status; s {
+		case "STOPPED":
+			return nil
+		case "RUNNING":
+			err = stopProgram(config, stopAddr)
+			if err != nil {
+				return err
+			}
+		case "STOPPING":
+			log.Println("program still in STOPPING state, waiting 10 seconds.")
+			time.Sleep(10 * time.Second)
+		default:
+			return fmt.Errorf("cannot stop program in app %s in state: %s", d.Get("app"), p.Status)
+		}
+	}
+}
+
+func resourceStreamingProgramExists(d *schema.ResourceData, m interface{}) (bool, error) {
+	config := m.(*Config)
+
+	addr := urlJoin(
+		config.host,
+		"/v3/namespaces", d.Get("namespace").(string),
+		"/apps", d.Get("app").(string), d.Get("type").(string),
+		d.Get("name").(string), "/status")
+
+	req, err := http.NewRequest(http.MethodGet, addr, nil)
+	if err != nil {
+		return false, err
+	}
+
+	b, err := httpCall(config.httpClient, req)
+	if err != nil {
+		return false, err
+	}
+
+	var p ProgramStatus
+	if err := json.Unmarshal(b, &p); err != nil {
+		return false, err
+	}
+
+	if p.Status == "RUNNING" {
```

Comment:
Exists should return if the resource itself exists, whether it's running or not shouldn't matter.

Reply:
So I think the way this Lifecycle API is structured makes this very confusing.
A program is not a first class resource in the CDAP API.
In other words, you cannot do a `GET ${host}/api/v3/namespaces/${namespace}/apps/${app}/spark` it will throw a `Not Found`

To list programs in an app you have to GET the app resource and read the array in the `programs` key
But this does not _really_ mean that this program exists.
evidence: you can create an app never start it and all the programs will be in the `programs` key

What we're trying to do with this resource is manage a program run (not something that CDAP makes easy for us).
The program run does not exist unless this status exists.

I'm trying to replicate behavior similar to [`dataflow_job`](https://www.terraform.io/docs/providers/google/r/dataflow_job.html)